### DISCUSSION TITLE
Fix Go the versions used by our workflow tests

### DISCUSSION THREAD

**Discussion by brackendawson** [stretchr/testify#1564] [Ideas]
# Current tests
We currently run full tests ("build") against:
- stable (1.22 today)
- oldstable (1.21 today)

Then partiual tests ("test") without formatting (because gofmt changed in 1.19) against:
- 1.17
- 1.18
- 1.19
- 1.20
- 1.21

It would be nice to not have to keep updating [this file](https://github.com/stretchr/testify/blob/bb548d0473d4e1c9b7bbfd6602c7bf12f7a84dd2/.github/workflows/main.yml) with new  Go versions, because human processes fail.

# Proposed
I think we can use this set long term:
- build: stable (because it is the latest supported Go release).
- build: oldstable (because it is the other supported Go release).
- build: 1.21 (because it is the oldest Go release which will fail to compile if a dependency has a go.mod go directive newer than the toolchain).
- test: [module](https://github.com/actions/setup-go?tab=readme-ov-file#getting-go-version-from-the-gomod-file) (currently 1.17 because it is the version we declare support for in our go.mod, we then don't have to have the comment in go.mod).

**Comment by ccoVeille** [comment]
I'm using @arnested go version action on multiple projects and I'm happy with it.

https://github.com/arnested/go-version-action

It solves the issue very conveniently

**Reply by dolmen** [reply]
This clearly nicely fits the need.

However for a critical project like Testify I want us to be very careful about supply chain security, so I prefer to avoid to depend on a non-official Action.

**Comment by dolmen** [comment]
We are using `stable`, `oldstable` since c86e13951134c5ddd6a9aa0d9ef9ec676e414893, merged in #1450.
